cmd/vicadmin/logs.go

```diff
@@ -298,3 +300,57 @@ func zipEntries(readers map[string]entryReader, out *zip.Writer) error {
 	}
 	return nil
 }
+
+func tailFile(wr io.Writer, file string, done *chan bool) error {
+	nlines := tailLines
+	f, err := os.Open(file)
+	if err != nil {
+		return err
+	}
+
+	pos, err := f.Seek(0, 2)
+	if err != nil {
+		return err
+	}
+	b := make([]byte, 1)
+	for pos > 0 && nlines >= 0 {
+		pos--
+		_, err := f.ReadAt(b, pos)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if b[0] == '\n' {
+			nlines--
+		}
+	}
+
+	pos++
+	if _, err = f.Seek(pos, 0); err != nil {
+		return err
+	}
+
+	b = make([]byte, 1024)
+	// We KNOW there's a data race here.
+	// But it doesn't break anything, so we just trap it.
+	defer func() {
+		_ = recover()
+	}()
+	var n int
+	for true {
+		for err != io.EOF {
+			n, err = f.Read(b)
+			if err != nil && err != io.EOF {
+				return err
+			}
+			fmt.Fprint(wr, string(b[:n]))
+		}
+		select {
+		case _ = <-*done:
```

Comment:
This is explicitly what the context package is for - that allows for explicit Cancellation, timeouts etc.